### ISSUE TITLE
Async/await guide

### ISSUE THREAD

**Issue by JulioJu** [Automattic/mongoose#7506]
Hi @libook  @mhombach  @vkarpov15 

Following the issue https://github.com/Automattic/mongoose/issues/7461 I would propose to improve API documentation.

For instance, for https://mongoosejs.com/docs/api.html#model_Model.findOneAndUpdate 

[Note: I've deleted a proposal here, because it was so bad]

Or instead as `A query is not a fully-fledged promise` (see https://mongoosejs.com/docs/promises.html) we could say instead:

    Returns
        Query (a thenable object)

Like that a user that know what is a Promise and a thenable and what is a callback could know easy how to use this function. When I've started with Mongoose, I thought falsely « dash, mongoose API doesn't return Promise contrary to the mongodb native driver ».

It should be done for all CRUD operations, into [lib/model.js](../tree/master/lib/model.js) and [lib/query.js](../tree/master/lib/query.js). 

Furthermore https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/mongoose/index.d.ts should be updated accordingly.  At least for this TypeScript definition file, we could say that: « findOneAndUpdate and others CRUD operations are Promises ». 

What do you think about that?

**Comment by JulioJu** [comment]
Probably you don't like document like it as `A query is not a fully-fledged promise` , but at least for TypeScript users the file  https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/mongoose/index.d.ts  could has this approximation.

**Comment by JulioJu** [comment]
Also, the example could be updated to:

```
A.findOneAndUpdate(conditions, update, options, callback) // executes
A.findOneAndUpdate(conditions, update, options)  // returns Query, a thenable object
A.findOneAndUpdate(conditions, update, callback) // executes
A.findOneAndUpdate(conditions, update)           // returns Query, a thenable object
A.findOneAndUpdate()                             // returns Query, a thenable object
```

**Comment by JulioJu** [comment]
Please check https://github.com/DefinitelyTyped/DefinitelyTyped/pull/32874

**Comment by vkarpov15** [comment]
There is some nuance that prevents us from having queries inherit from promises. See https://mongoosejs.com/docs/queries.html#queries-are-not-promises for more details. With #7398 we should be able to make queries extend from promises.

I'm not sure I like adding parentheses after queries return value. Maybe we add examples of using async/await with queries to the API docs instead?

**Comment by JulioJu** [comment]
Ok I see, 

But for https://github.com/DefinitelyTyped/DefinitelyTyped/pull/32874 the problem is that if Query doesn't inherits from Promises, `tslint` gives errors `https://github.com/DefinitelyTyped/DefinitelyTyped/pull/32874 [await-promise](https://palantir.github.io/tslint/rules/await-promise/) and [no-invalid-await](https://github.com/SonarSource/SonarTS/blob/master/sonarts-core/docs/rules/no-invalid-await.md) . 

If you say that it's not a good idea to say in the TypeScript definition file that Query inherits from Promises, maybe in the doc we could inform that tslint errors `await-promise` and `no-invalid-await` are false positive?

**Comment by libook** [comment]
Hi @JulioJu, thanks for your concern for my issue.

As we know, 'Queries' are not Promises but 'thenable' objects.
And the await command is not only for Promises. Actually, await command works with thenable objects.

```javascript
//This will return a thenable object.
function foo(x) {
  return {
    then: function(resolve, reject) {
      setTimeout(resolve, 3000, 'The x is: ' + x);
      return this;
    }
  };
}

(async () => await foo(2))().then(console.log);//Use await to execute the thenable object.
```

I guess this is how Queries support await command.

So, what if this is a issue of tslint? Checking whether the object has a then method, instead of checking whether the object is a Promise.

I don't use TypeScript and tslint. So there may be something I don't know. I'll be happy if you point them out.

**Comment by JulioJu** [comment]
Sorry, but your sample is wrong,

Try: 
```javascript
//This will return a thenable object.
function foo(x) {
  return {
    then: function(resolve, reject) {
      setTimeout(resolve, 3000, 'The x is: ' + x);
      return this;
    }
  };
}

(() => Promise.resolve(foo(2)))().then(console.log);
```

If I understand well javascript, `(async () => await foo(2)) === (async () => Promise.resolve(foo(2))) === (() => Promise.resolve(foo(2)))`

Furthermore 

> The await expression causes async function execution to pause until a Promise is resolved, that is fulfilled or rejected, and to resume execution of the async function after fulfillment. (…) If the value of the expression following the await operator is not a Promise, it's converted to a resolved Promise.

https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/await

**Comment by JulioJu** [comment]
Maybe a better example is the following. In the following case, `async` has a true sens. 
```javascript
            function foo(x) {
                return {
                    then: function(resolve, reject) {
                        setTimeout(resolve, 3000, 'The x is: ' + x);
                    return this;
                    }
                };
            }

            (async () => {const x = await foo(2) ; console.log(x) ; })().then(console.log('coucou'));
```

**Comment by JulioJu** [comment]
I am not going to post something in `tslint`. It's a tricky side effect, probably more complicate to lint. Actually, probably they detect correctness of `await` use simply to check if the expression is `instenceof` Promise. `await` was disigned for that

> The await expression causes async function execution to pause until a Promise is resolved, 
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/await

I believe it's not the goal of a linter to check non standard uses as described above. `await` behaviour is deeply documented for promises, and probably think for that. 

But I continue to think that the approximation presented in [this link](https://github.com/DefinitelyTyped/DefinitelyTyped/pull/32874/commits/4826667c6017adbd94086e9875fbc2abf0ce6aba) is a good very solution for the TypeScript definition file. Otherwise people could write without warnings

```
myMongooseService = () => {
    const myResult = Query(2);
    console.log('I expect the result: ', x);
    return myResult;
}
```

**Comment by JulioJu** [comment]
Therefore, for https://github.com/DefinitelyTyped/DefinitelyTyped/pull/32874 all it's ok, I've tested ;-). 

For the doc https://mongoosejs.com/docs/api.html I propose to comment like following: 

```
Returns
    Query, a thenable object

Example
A.findOneAndUpdate(conditions, update, options, callback) // returns Query and executes the request
A.findOneAndUpdate(conditions, update, options)  // returns Query
A.findOneAndUpdate(conditions, update, callback) // returns Query and executes the request
A.findOneAndUpdate(conditions, update)           // returns Query
A.findOneAndUpdate()                             // returns Query
```

Maybe somwhere in the doc, https://github.com/Automattic/mongoose/issues/7506#issuecomment-463978475 could be added to explain how Mongoose work with `await / async`.

**Comment by vkarpov15** [comment]
@JulioJu adding an async/await guide is a good idea. There's a lot of nuance with using async/await, that's why [I wrote a book about it](http://asyncawait.net/). We'll also add a note that queries are thenable to the API docs

**Comment by Rossh87** [comment]
I'd be happy to write this if it's still an issue of interest.  Would you like me to fork and create a new .pug file, or edit it into an existing file?

**Comment by vkarpov15** [comment]
@Rossh87 sure, new `.pug` file would be great. The docs are a bit tricky to work with, but running `node ./static.js` to start a static server and `make gendocs` to compile the docs will get you most of the way there.

**Comment by Rossh87** [comment]
@vkarpov15 I've got a PR ready to go that includes the new `docs/async-await.pug`, as well as changes to `docs/layout.pug` and `source/index.js` that are needed to get the new file added to site navigation.  However, said PR also includes the newly-generated docs from running `gendocs`.  From the contributing guidelines it looks like that may not be what you prefer; if not, let me know and I can get a commit set up with just the relevant file changes and no new documentation.  Otherwise I'll submit what I've got.
